File: CSharp/NullObject/NullObjectWithIoC/BankAccount.cs
namespace NullObjectWithIoC
{
    public class BankAccount
    {
        private readonly ILog _log;
        private int balance;

        public BankAccount(ILog log)
        {
            _log = log;
        }

        public void Deposit(int amount)
        {
            balance += amount;
            _log?.Info($"Deposited {amount}, balance is now {balance}");
        }
    }
}

File: Python/Memento/MementoBankAccount/bank_account.py
from memento import Memento


class BankAccount:
    def __init__(self, balance: int = 0) -> None:
        self.balance = balance

    def deposit(self, amount: int) -> Memento:
        self.balance += amount
        return Memento(self.balance)

    def restore(self, memento: Memento) -> None:
        self.balance = memento.balance

    def __str__(self) -> str:
        return f"Balance = {self.balance}"
